import os
from pathlib import Path

configfile: "../config/config.json"

with open("SRR_id.txt") as f:
    id_list = f.read().splitlines()

include: "rules/download_all.smk"
include: "rules/common_seq.smk"
include: "rules/dge_analysis.smk"

rule all:
    input:
        qc_fq1_out = expand("data/qc/{id}_1_fastqc.html", id = id_list),
        trimm_fq1 = expand("data/trimmed/{id}_1.fastq.gz", id = id_list),
        trimm_fq2 = expand("data/trimmed/{id}_2.fastq.gz", id = id_list),
        qc_trimm_fq2_out = expand("data/qc_trimmed/{id}_2_fastqc.html", id = id_list),
        chrNameLength = config['path']['chrNameLength'],
        bam = expand("results/STAR/{id}/Aligned.sortedByCoord.out.bam", id = id_list),
        bam_logs = expand("results/STAR/{id}/Log.final.out", id = id_list)

rule all_downloads:
    input:
        gtf = expand('data/references/gtf/homo_sapiens.gtf', **config),
        genome = expand('data/references/genome_fa/homo_sapiens_genome.fa'),
        fake_output_fastq = expand('data/references/geo_download.txt')

rule dge:
    input:
        counts = expand("results/featureCounts/{id}_gene_counts.tsv", id = id_list)
