import os
from pathlib import Path
import pandas as pd

configfile: "../config/config.json"

with open("SRR_id.txt") as f:
    id_list = f.read().splitlines()

include: "rules/download_all.smk"
include: "rules/common_seq.smk"
include: "rules/dge_analysis.smk"


rule all:
    input:
        qc_fq1_out = expand("data/qc/{id}_1_fastqc.html", id = id_list),
        trimm_fq1 = expand("data/trimmed/{id}_1.fastq.gz", id = id_list),
        trimm_fq2 = expand("data/trimmed/{id}_2.fastq.gz", id = id_list),
        qc_trimm_fq2_out = expand("data/qc_trimmed/{id}_2_fastqc.html", id = id_list),
        chrNameLength = config['path']['chrNameLength'],
        bam = expand("results/STAR/{id}/Aligned.sortedByCoord.out.bam", id = id_list),
        bam_logs = expand("results/STAR/{id}/Log.final.out", id = id_list)


rule all_downloads:
    input:
#        fake_output_fastq = expand('data/references/geo_download.txt'),
        coco_dependency_download = "scripts/pairedBamToBed12/",
        coco_git = 'git_repos/coco'


rule dge:
    input:
        transcriptome = "data/references/transcriptome.fa",
        kallisto_index = "data/references/kallisto.idx",
        kallisto_quant = expand("results/dge/kallisto/{id}/abundance.tsv", id = id_list),
        tx2gene = "data/references/tx2gene.tsv",
        pc_gtf = "data/references/gtf/hg38_Ensembl_V101_Scottlab_2020.protein_coding.gtf",
        tpm = "results/dge/kallisto/tpm.tsv",
        deseq2 = "results/dge/deseq2/FXS-Control.csv"
